ocaml/xapi/xapi_pool_update.ml

```diff
@@ -480,6 +480,25 @@ let resync_host ~__context ~host =
     (* Clean updates that don't have a corresponding patch record *)
     Db.Pool_update.get_all ~__context
     |> List.filter (fun self -> Xapi_pool_patch.pool_patch_of_update ~__context self = Ref.null)
+    |> List.iter (fun self -> destroy ~__context ~self);
+
+    (*
+     * If db indicates an update is not applied to any host but the corresponding patch is applied
+     * on some host(s), that means the RPU has completed, the update record should be removed.
+     * (The patch record will be removed along with the update record.)
+    *)
+    Db.Pool_update.get_all ~__context
+    |> List.filter (fun self ->
+                      if List.length (Db.Pool_update.get_hosts ~__context ~self) = 0 then begin
+                        let patch = Xapi_pool_patch.pool_patch_of_update ~__context self in
+                        if List.length (Db.Pool_patch.get_host_patches ~__context ~self:patch) = 0 then
```

Comment:
This can be shortened to `Db.Pool_patch.get_host_patches ~__context ~self:patch <> []` as it will yield `false` on an empty list and true otherwise. Another option is:

```
Xapi_pool_patch.pool_patch_of_update ~__context self
|> fun self -> Db.Pool_patch.get_host_patches ~__context ~self
|> match [] -> false | _ -> true
```